File: LeetCode/C/543-tree-二叉树的直径.cpp
/*
 * @lc app=leetcode.cn id=543 lang=cpp
 *
 * [543] 二叉树的直径
 */
#include "main.h"
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
  public:
    int diameterOfBinaryTree(TreeNode *root) {
        int res = 1;
        dfs(root, res);
        return res - 1;
    }

  private:
    int dfs(TreeNode *root, int &res) {
        if (root == nullptr)
            return 0;
        int l = dfs(root->left, res);
        int r = dfs(root->right, res);
        res = max(res, l + r + 1);
        return max(l, r) + 1;
    }
};
// @lc code=end

File: LeetCode/Java/543-tree-二叉树的直径.java
/*
 * @lc app=leetcode.cn id=543 lang=java
 *
 * [543] 二叉树的直径
 */

// @lc code=start
class Solution {
    private int ans;

    public int diameterOfBinaryTree(TreeNode root) {
        ans = 1;
        depth(root);
        return ans - 1;
    }

    private int depth(TreeNode node) {
        if (node == null)
            return 0;
        int l = depth(node.left);
        int r = depth(node.right);
        ans = Math.max(ans, l + r + 1);
        return Math.max(l, r) + 1;
    }
}
// @lc code=end
